optimum/habana/trl/trainer/ddpo_trainer.py

```diff
@@ -0,0 +1,506 @@
+        - Reduce recompilations by avoiding constant variables in loops
+        - Add `mark_step()` to support lazy mode
+
+        Args:
+            inner_epoch (int): The current inner epoch
+            epoch (int): The current epoch
+            global_step (int): The current global step
+            batched_samples (List[Dict[str, torch.Tensor]]): The batched samples to train on
+
+        Side Effects:
+            - Model weights are updated
+            - Logs the statistics to the accelerator trackers.
+
+        Returns:
+            global_step (int): The updated global step
+        """
+        info = defaultdict(list)
+
+        for i, sample in enumerate(batched_samples):
+            if self.config.train_cfg:
+                # concat negative prompts to sample prompts to avoid two forward passes
+                embeds = torch.cat([sample["negative_prompt_embeds"], sample["prompt_embeds"]])
+            else:
+                embeds = sample["prompt_embeds"]
+
+            latents = sample["latents"]
+            timesteps = sample["timesteps"]
+            next_latents = sample["next_latents"]
+            log_probs = sample["log_probs"]
+
+            for j in range(self.num_train_timesteps): #, desc=f"Epoch{i}"):
+                with self.accelerator.accumulate(self.sd_pipeline.unet):
+                    # Reduce recompilations by avoiding constant variables in loops
+                    latent = latents[:, 0]
+                    timestep = timesteps[:, 0]
+                    next_latent = next_latents[:, 0]
+                    log_prob = log_probs[:, 0]
+                    latents = torch.roll(latents, shifts=-1, dims=1)
+                    timesteps = torch.roll(timesteps, shifts=-1, dims=1)
+                    next_latents = torch.roll(next_latents, shifts=-1, dims=1)
+                    log_probs = torch.roll(log_probs, shifts=-1, dims=1)
+
+                    loss, approx_kl, clipfrac = self.calculate_loss(
+                        latent,
+                        timestep,
+                        next_latent,
+                        log_prob,
+                        sample["advantages"],
+                        embeds,
+                    )
+
+                    info["approx_kl"].append(approx_kl)
+                    info["clipfrac"].append(clipfrac)
+                    info["loss"].append(loss)
+
+                    self.accelerator.backward(loss)
+                    if self.accelerator.sync_gradients:
+                        self.accelerator.clip_grad_norm_(
+                            self.trainable_layers.parameters()
+                            if not isinstance(self.trainable_layers, list)
+                            else self.trainable_layers,
+                            self.config.train_max_grad_norm,
+                        )
+                    self.optimizer.step()
+                    self.optimizer.zero_grad()
+                    if self.use_habana and not self.use_hpu_graphs:
```

Comment:
can you please explain why we can't run mark_step if use_hpu_graphs is True?

Reply:
I'll add a README with an example. 

For mark_step, we were following the implementation here https://github.com/huggingface/optimum-habana/blob/2f55de3f9ad23a853768c16029c455ee11b0e515/optimum/habana/diffusers/pipelines/stable_diffusion/pipeline_stable_diffusion.py#L534 

I can delete the condition in our implementation

Reply:
Got it. can you please hold this change then? let me check further.

Reply:
@yeonsily I removed the hpu graph condition based on your guidance and https://docs.habana.ai/en/latest/PyTorch/Model_Optimization_PyTorch/HPU_Graphs_Training.html